turbo/snapshotsync/freezeblocks/caplin_snapshots.go

```diff
@@ -280,21 +282,65 @@ Loop:
 		s.segmentsMax.Store(segmentsMax)
 	}
 	s.segmentsReady.Store(true)
+
+	s.calculateVisibleSegments()
 	s.idxMax.Store(s.idxAvailability())
 	s.indicesReady.Store(true)
 
 	return nil
 }
 
-func (s *CaplinSnapshots) idxAvailability() uint64 {
-	var beaconBlocks uint64
-	for _, seg := range s.BeaconBlocks.segments {
-		if seg.Index() == nil {
+func (s *CaplinSnapshots) calculateVisibleSegments() {
+	var maxIndex = -1
+	var maxBlockNum uint64
+	if len(s.BeaconBlocks.dirtySegments) == 0 {
+		return
+	}
+	slices.SortFunc(s.BeaconBlocks.dirtySegments, func(i, j *Segment) int {
```

Comment:
“CalcVisibleFiles”: must not mutate dirty files. It’s not expected side-effect from func name.

Reply:
Update the func name to `divideSegmentsIntoDirtyAndVisible`. 

Every time `rebuildSegments` is called, only dirty files are reopened and rebuild. After rebuilding dirty segments, this func will calculate new visible segments,  append them to existing visible segments and delete from dirty segments.  
Visible segments are created/indexed/seedable(won't be merged into bigger files). These properties make visible segments append-only, so we don't have to re-calculate visible segments during rebuild.

Reply:
1. “ Visible segments are created/indexed/seedable(won't be merged into bigger files)” - not true - small files also can be visible and merge when have enough small files. 
2. If append something to existing slice: existing users will see this changes. Means no transactions-isolation for existing readers.

Reply:
I understand that in `aggregator.go` small files can be visible before merge, and that's exactly what I want to discuss with you.

I change the definition of visible here a bit to make sure that once a file is add to visible, it won't be updated again.  This simplifies the maintaining of visible files - don't need an extra mechanism(`refcount`) to decide when to delete the small files. However, the side-effect is DB cannot prune blocks in small files until they are merged into bigger one.  

I think you must have thought about this and want to know your consideration.

Reply:
I see - unfortunately: 
- E3 does need keep DB as small as possible. Means prune as early as possible. It’s part of performance.
- files which never merged to bigger one - we call “frozen”.

Reply:
Sorry for late reply.  
I updated the PR these days, and now the design of dirty/visible segments is more similar to dirty/visible files used in `aggregator.go`.